python/dgl/nn/pytorch/explain/gnnexplainer.py

```diff
@@ -425,3 +428,621 @@ def explain_graph(self, graph, feat, **kwargs):
+    num_epochs : int, optional
+        The number of epochs to train.
+    alpha1 : float, optional
+        A higher value will make the explanation edge masks more sparse by decreasing
+        the sum of the edge mask.
+    alpha2 : float, optional
+        A higher value will make the explanation edge masks more sparse by decreasing
+        the entropy of the edge mask.
+    beta1 : float, optional
+        A higher value will make the explanation node feature masks more sparse by
+        decreasing the mean of the node feature mask.
+    beta2 : float, optional
+        A higher value will make the explanation node feature masks more sparse by
+        decreasing the entropy of the node feature mask.
+    log : bool, optional
+        If True, it will log the computation process, default to True.
+    """
+
+    def __init__(self,
+                 model,
+                 num_hops,
+                 lr=0.01,
+                 num_epochs=100,
+                 *,
+                 alpha1=0.005,
+                 alpha2=1.0,
+                 beta1=1.0,
+                 beta2=0.1,
+                 log=True):
+        super(HeteroGNNExplainer, self).__init__()
+        self.model = model
+        self.num_hops = num_hops
+        self.lr = lr
+        self.num_epochs = num_epochs
+        self.alpha1 = alpha1
+        self.alpha2 = alpha2
+        self.beta1 = beta1
+        self.beta2 = beta2
+        self.log = log
+
+    def _init_masks(self, graph, feat):
+        r"""Initialize learnable feature and edge mask.
+
+        Parameters
+        ----------
+        graph : DGLGraph
+            Input graph.
+        feat : dict[str, Tensor]
+            The dictionary that associates input node features(values) with
+            the respective node types(keys) present in the graph.
+
+        Returns
+        -------
+        feat_masks : dict[str, Tensor]
+            Dictionary of { ntypes: features }
+            The dictionary that associates the feature masks(values) with
+            the respective node types(keys).
+            Feature mask of shape :math:`(1, D_t)`, where :math:`D_t`
+            is the feature size for node type :math:`t`.
+        edge_masks : dict[tuple[str], Tensor]
+            Dictionary of { canonical_etypes: features }
+            The dictionary that associates the edge masks(values) with
+            the respective canonical edge types(keys).
+            Edge mask of shape :math:`(E)`, where :math:`E` is the
+            number of edges.
+        """
+        device = graph.device
+        feat_masks = {}
+        for node_type, feature in feat.items():
+            std = 0.1
+            num_nodes, feat_size = feature.size()
+            feat_masks[node_type] = nn.Parameter(torch.randn(1, feat_size, device=device) * std)
+
+        edge_masks = {}
+        for canonical_etype in graph.canonical_etypes:
+            src = canonical_etype[0]
+            num_nodes = graph.num_nodes(src)
```

Comment:
Is there a particular reason for using `graph.num_nodes(canonical_etype[0])` over `graph.num_nodes(canonical_etype[-1])`?